rfcs/2022-03-15-11851-ingest-opentelemtry-traces.md

```diff
@@ -0,0 +1,384 @@
+is rather verbose and comes with complex nested field. The [Datadog approach][span-conversion] is either to ignore those
+(e.g. the links field is ignored) or encode the complete field into a text representation (e.g. events are encoded using
+JSON) and include the resulting value into the tags (a.k.a Meta) map.
+
+This makes the opposite conversion a bit complicated if we want it to be completely symetrical but there was already an
+[attempt][otlp-dd-trace-receiver] allow Datadog traces ingestion in the OpenTelemetry collector. While this PR was
+closed unmerged this provide a valuable example. Anyways the [otlp-and-other-formats][OpenTelemetry] acknowledges that
+some of the OpenTelemetry contruct ends up being stored as tags or annotations in other formats.
+
+Anyway the OpenTelemtry to Datadog traces conversion is dictacted by existing implementation in both the `trace-agent`
+and the Datadog exporter as users will expect a consistent behaviour from one solution to another. The same
+consideration applies for APM stats computation, as [official implementations][apm-stats-computation] already provides a
+reference that define what should be done to get the same result with Vector in the loop. The other way, from Datadog to
+OpenTelemetry is less common as of today but while implementing conversions we should ensure that the following path is
+at best idempotent but at least fully functional:
+
+`(Datadog Trace) -> (Vector internal format - based on Opentelemetry) -> (Datadog Trace)`
+
+**Note**: The [Rust OpenTelemetry implementation][otlp-rust] implement a conversion from OpenTelemetry traces to the
+Datadog `trace-agent` format. This is not the purpose of this RFC, and with the OpenTelemetry traces format being
+supported on both sides working on better interoperability on that particular common ground would likely be a better
+option.
+
+**Conclusion**: the implementation will stay around [./lib/vector-core/src/event/trace.rs][current-trace-in-vector], it
+will borrow most of the OpenTelemetry to allow straightforward trace conversion to the newer Vector internal
+representation. Regarding `datadog_agent` source and `datadog_traces` sink the conversion to/from this newer trace representation will follow existing logic and ensure that standard usecases (like introducing Vector between the Datadog intake and the `trace agen`) do not signigicantly change the end-to-end behaviour. Some top-level information (Like trace ID, trace-wide tags/metrics, the original format) are likely to be added to the internal trace representation for efficiency and convenience.
+Trace would not get native `VrlTarget` representation anymore, there is a bigger discussion there that should probably
+be adressed separately. As an interim measure few fields may be exposed (At least trace ID & trace-wide tags), the spans
+list will not be exposed initially.
+
+#### APM stats computation
```

Comment:
Have you considered the alternative of, instead of having the sources do these calculations of APM, have them done in the `datadog_traces` sink? I'm wondering:

* If the metrics are useful outside of the context of Datadog. A naive reading of them makes it seem like this in a Datadog specific enhancement.
* If users would do transformations that would affect these metrics. For example, by filtering out traces. If a user filters out traces, would we still want the APM metrics to reflect all of them?

Reply:
After thinking about this some more. I am actually wondering if this isn't better as a separate transform to give users more control over when this calculation happens. For example, maybe they'd like to filter before calculating the stats? Maybe I'm investing non-existent use cases though 😄 

What if we started with a transform and then added it to the trace sources later? They could just integrate it by running the transform directly in the source.

Reply:
This is IMHO a relevant use case (maybe not immediately). If you imagine a trace transform that would limit the depth of a trace by removing span deeper than a certain threshold, computing APM stats after this transform is a good idea as it would prevent APM stats to be computed for dropped spans. Also in case of heavy tag manipulation it's probably safe to assume that doing that once and then compute APM stats is probably better than doing the same operation on both the trace flow and the APM stats flow.

Implementation-wise this is probably a good idea to keep that in mind and make it flexible enough (in the end this will always be calculating APM stats on traces using the internal vector representation) so we should be able to implement both with the same underlying implementation and we may ultimately keep both option: e.g. a pass-through configuration would enabled APM stats computation in the source directly connected to a sink while a more complex configuration with routing, tag manipulation, etc. and get the APM computed at the most efficient place.